### DISCUSSION TITLE
Softplus as a better alternative to Exponential function for Positivity Constraint

### DISCUSSION THREAD

**Discussion by AhmedThahir** [OTexts/Discussion#3] [Q&A] [answered]
Hi moderators,

I was reading the ebook fpp3. This is a doubt pertaining to the section: https://otexts.com/fpp3/limits.html#positive-forecasts.

I would like to know what is your opinion on using the Softplus function instead of an exponential function for ensuring positive forecasts?

![image](https://github.com/user-attachments/assets/e120c521-9311-4d1e-bc24-0d7b61de3c9b)

From my understanding, Softplus has desirable properties:
- Softplus converges to linear function when x is positive
- Softplus converges to exp function when x is negative

Looking forward to hearing your analysis.

**Comment by robjhyndman** [comment, marked as answer]
Yes, softplus is an alternative transformation that can be used to ensure positive forecasts. The advantage of exponential is that it allows for multiplicative interpretations. But sometimes the exponential is too strong, and then softplus can be useful.

Here is an example:

``` r
library(fable)

softplus <- function(x) {
    log(1 + exp(x))
}
invsoftplus <- function(x) {
    log(exp(x) - 1)
}

inv_softplus <- new_transformation(softplus, invsoftplus)

USAccDeaths |>
    as_tsibble() |>
    model(arima = ARIMA(inv_softplus(value/1e3))) |>
    forecast()
#> # A fable: 24 x 4 [1M]
#> # Key:     .model [1]
#>    .model    index           value  .mean
#>    <chr>     <mth>          <dist>  <dbl>
#>  1 arima  1979 Jan  t(N(8.3, 0.1))  8336.
#>  2 arima  1979 Feb t(N(7.5, 0.14))  7532.
#>  3 arima  1979 Mar t(N(8.3, 0.17))  8315.
#>  4 arima  1979 Apr  t(N(8.6, 0.2))  8617.
#>  5 arima  1979 May t(N(9.5, 0.24))  9489.
#>  6 arima  1979 Jun t(N(9.9, 0.27))  9860.
#>  7 arima  1979 Jul   t(N(11, 0.3)) 10907.
#>  8 arima  1979 Aug  t(N(10, 0.34)) 10086.
#>  9 arima  1979 Sep t(N(9.2, 0.37))  9165.
#> 10 arima  1979 Oct  t(N(9.4, 0.4))  9384.
#> # ℹ 14 more rows
```

<sup>Created on 2024-12-08 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Comment by AhmedThahir** [comment]
> The advantage of exponential is that it allows for multiplicative interpretations.

What would be the interpretation of coefficients with a softplus transformation?

**Comment by robjhyndman** [comment]
There is no simple interpretation

**Reply by AhmedThahir** [reply]
I have a hypothesis.

Since
1. We are modelling a variable which only takes non-negative values
2. Softplus ≈ Linear for non-negative values

Linear regression and softplus regression would give the same coefficients.

Hence, the interpretation of softplus regression would be the same as that of linear regression coefficients. 

Pls evaluate my hypothesis, and advice if it would be useful to validate it experimentally.

**Reply by robjhyndman** [reply]
softplus is not linear for non-negative values. Look at the graph. It is only asymptotically linear.

**Reply by AhmedThahir** [reply]
We could use a generalized softplus function

$$
g(x) = \dfrac{\ln \vert 1 + \exp(kx) \vert}{k}
$$

- k = 1 is the standard softplus
- As k gets larger, the softplus gets closer to the linear function for the non-negative inputs

![Screenshot_20241231-223515_Brave](https://github.com/user-attachments/assets/05f73be6-4e89-436b-b5ac-6410514aaaec)

![Screenshot_20241231-223532_Brave](https://github.com/user-attachments/assets/e7075331-bda2-4efa-9744-a1b6c4f68de5)

This is natively supported in [PyTorch](https://pytorch.org/docs/stable/generated/torch.nn.Softplus.html)
